File: packages/contracts-ts/src/generation.ts
import { z } from 'zod';

import {
  correlationIdSchema,
  generationRequestIdSchema,
  traceIdSchema,
  utcTimestampSchema,
} from './primitives';
import {
  GENERATION_ARTIFACT_SCHEMA_VERSION,
  GENERATION_REQUEST_SCHEMA_VERSION,
  GENERATION_RESPONSE_SCHEMA_VERSION,
  MODEL_CONFIG_SCHEMA_VERSION,
  PROVIDER_CAPABILITIES_SCHEMA_VERSION,
  PROVIDER_ERROR_SCHEMA_VERSION,
  PROVIDER_GENERATE_REQUEST_SCHEMA_VERSION,
  PROVIDER_GENERATE_RESPONSE_SCHEMA_VERSION,
  PROVIDER_USAGE_SCHEMA_VERSION,
  RECORDED_RESPONSE_KEY_SCHEMA_VERSION,
  STRUCTURED_OUTPUT_SPEC_SCHEMA_VERSION,
  TOOL_SCHEMA_SCHEMA_VERSION,
} from './versioning';

const schemaVersionCheck = (expected: number) =>
  z
    .number()
    .int()
    .min(1)
    .refine((v) => v === expected);

export const GenerationMessageRole = {
  SYSTEM: 'system',
  USER: 'user',
  ASSISTANT: 'assistant',
  TOOL: 'tool',
} as const;

export const ProviderCapability = {
  STRUCTURED_OUTPUT: 'structured_output',
  TOOLS: 'tools',
  VISION: 'vision',
  STREAMING: 'streaming',
} as const;

export const ProviderFinishReason = {
  STOP: 'stop',
  LENGTH: 'length',
  TOOL_CALLS: 'tool_calls',
  CONTENT_FILTER: 'content_filter',
  ERROR: 'error',
} as const;

export const ProviderErrorCode = {
  VALIDATION_FAILED: 'VALIDATION_FAILED',
  CAPABILITY_UNSUPPORTED: 'CAPABILITY_UNSUPPORTED',
  CREDENTIALS_MISSING: 'CREDENTIALS_MISSING',
  OUTPUT_LIMIT_EXCEEDED: 'OUTPUT_LIMIT_EXCEEDED',
  STRUCTURED_OUTPUT_INVALID: 'STRUCTURED_OUTPUT_INVALID',
  TIMEOUT: 'TIMEOUT',
  RETRY_EXHAUSTED: 'RETRY_EXHAUSTED',
  CIRCUIT_OPEN: 'CIRCUIT_OPEN',
  PROVIDER_UNAVAILABLE: 'PROVIDER_UNAVAILABLE',
  CANCELLATION: 'CANCELLATION',
  INTERNAL: 'INTERNAL',
} as const;

export const generationMessageSchema = z.object({
  role: z.enum([
    GenerationMessageRole.SYSTEM,
    GenerationMessageRole.USER,
    GenerationMessageRole.ASSISTANT,
    GenerationMessageRole.TOOL,
  ]),
  content: z.string().min(0).max(65536),
  name: z.string().max(128).optional(),
});

export const toolSchemaSchema = z.object({
  schemaVersion: schemaVersionCheck(TOOL_SCHEMA_SCHEMA_VERSION),
  name: z.string().min(1).max(128),
  description: z.string().min(1).max(2048),
  parameters: z.record(z.unknown()).default({}),
});

export const structuredOutputSpecSchema = z.object({
  schemaVersion: schemaVersionCheck(STRUCTURED_OUTPUT_SPEC_SCHEMA_VERSION),
  jsonSchema: z.record(z.unknown()),
  strict: z.boolean().default(true),
  maxRepairAttempts: z.number().int().min(0).max(2).default(1),
});

export const providerCapabilitiesSchema = z.object({
  schemaVersion: schemaVersionCheck(PROVIDER_CAPABILITIES_SCHEMA_VERSION),
  capabilities: z
    .array(
      z.enum([
        ProviderCapability.STRUCTURED_OUTPUT,
        ProviderCapability.TOOLS,
        ProviderCapability.VISION,
        ProviderCapability.STREAMING,
      ]),
    )
    .default([]),
});

export const modelConfigSchema = z.object({
  schemaVersion: schemaVersionCheck(MODEL_CONFIG_SCHEMA_VERSION),
  providerId: z.string().min(1).max(64),
  modelId: z.string().min(1).max(128),
  promptVersion: z.string().min(1).max(64),
  temperature: z.number().min(0).max(2).optional(),
  maxOutputTokens: z.number().int().min(1).max(65536).optional(),
});

export const providerUsageSchema = z
  .object({
    schemaVersion: schemaVersionCheck(PROVIDER_USAGE_SCHEMA_VERSION),
    promptTokens: z.number().int().min(0),
    completionTokens: z.number().int().min(0),
    totalTokens: z.number().int().min(0),
    estimatedCostUsd: z.number().min(0).optional(),
  })
  .superRefine((value, ctx) => {
    if (value.totalTokens !== value.promptTokens + value.completionTokens) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'totalTokens must equal promptTokens + completionTokens',
      });
    }
  });

export const providerErrorSchema = z.object({
  schemaVersion: schemaVersionCheck(PROVIDER_ERROR_SCHEMA_VERSION),
  code: z.enum([
    ProviderErrorCode.VALIDATION_FAILED,
    ProviderErrorCode.CAPABILITY_UNSUPPORTED,
    ProviderErrorCode.CREDENTIALS_MISSING,
    ProviderErrorCode.OUTPUT_LIMIT_EXCEEDED,
    ProviderErrorCode.STRUCTURED_OUTPUT_INVALID,
    ProviderErrorCode.TIMEOUT,
    ProviderErrorCode.RETRY_EXHAUSTED,
    ProviderErrorCode.CIRCUIT_OPEN,
    ProviderErrorCode.PROVIDER_UNAVAILABLE,
    ProviderErrorCode.CANCELLATION,
    ProviderErrorCode.INTERNAL,
  ]),
  message: z.string().min(1).max(2048),
  retryable: z.boolean().default(false),
  details: z.record(z.unknown()).default({}),
  traceId: traceIdSchema.optional(),
});

export const recordedResponseKeySchema = z.object({
  schemaVersion: schemaVersionCheck(RECORDED_RESPONSE_KEY_SCHEMA_VERSION),
  providerId: z.string().min(1).max(64),
  modelId: z.string().min(1).max(128),
  promptVersion: z.string().min(1).max(64),
  requestHash: z.string().min(8).max(128),
});

export const generationRequestSchema = z.object({
  schemaVersion: schemaVersionCheck(GENERATION_REQUEST_SCHEMA_VERSION),
  requestId: generationRequestIdSchema,
  traceId: traceIdSchema,
  correlationId: correlationIdSchema.optional(),
  providerId: z.string().max(64).optional(),
  modelConfig: modelConfigSchema,
  messages: z.array(generationMessageSchema).min(1).max(64),
  tools: z.array(toolSchemaSchema).max(32).default([]),
  structuredOutput: structuredOutputSpecSchema.optional(),
  capabilitiesRequired: z
    .array(
      z.enum([
        ProviderCapability.STRUCTURED_OUTPUT,
        ProviderCapability.TOOLS,
        ProviderCapability.VISION,
        ProviderCapability.STREAMING,
      ]),
    )
    .default([]),
  maxOutputTokens: z.number().int().min(1).max(65536).default(4096),
  timeoutMs: z.number().int().min(100).max(600000).optional(),
  idempotencyKey: z.string().max(256).optional(),
});

export const generationResponseSchema = z.object({
  schemaVersion: schemaVersionCheck(GENERATION_RESPONSE_SCHEMA_VERSION),
  requestId: generationRequestIdSchema,
  traceId: traceIdSchema,
  providerId: z.string().min(1).max(64),
  modelId: z.string().min(1).max(128),
  promptVersion: z.string().min(1).max(64),
  content: z.string().max(65536).optional(),
  structuredData: z.record(z.unknown()).optional(),
  finishReason: z.enum([
    ProviderFinishReason.STOP,
    ProviderFinishReason.LENGTH,
    ProviderFinishReason.TOOL_CALLS,
    ProviderFinishReason.CONTENT_FILTER,
    ProviderFinishReason.ERROR,
  ]),
  usage: providerUsageSchema,
  latencyMs: z.number().int().min(0),
  artifactRef: z.string().max(512).optional(),
  completedAt: utcTimestampSchema,
});

export const generationArtifactSchema = z.object({
  schemaVersion: schemaVersionCheck(GENERATION_ARTIFACT_SCHEMA_VERSION),
  requestId: generationRequestIdSchema,
  traceId: traceIdSchema,
  providerId: z.string().min(1).max(64),
  modelId: z.string().min(1).max(128),
  promptVersion: z.string().min(1).max(64),
  latencyMs: z.number().int().min(0),
  usage: providerUsageSchema.optional(),
  error: providerErrorSchema.optional(),
  sanitizedRequest: z.record(z.unknown()),
  sanitizedResponse: z.record(z.unknown()).optional(),
  objectStorageRef: z.string().max(512).optional(),
  recordedAt: utcTimestampSchema,
});

export const providerGenerateRequestSchema = z.object({
  schemaVersion: schemaVersionCheck(PROVIDER_GENERATE_REQUEST_SCHEMA_VERSION),
  request: generationRequestSchema,
  providerId: z.string().max(64).optional(),
  dryRun: z.boolean().default(false),
});

export const providerGenerateResponseSchema = z
  .object({
    schemaVersion: schemaVersionCheck(PROVIDER_GENERATE_RESPONSE_SCHEMA_VERSION),
    response: generationResponseSchema.optional(),
    error: providerErrorSchema.optional(),
  })
  .superRefine((value, ctx) => {
    if ((value.response === undefined) === (value.error === undefined)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Exactly one of response or error must be set',
      });
    }
  });

export type GenerationMessageV1 = z.infer<typeof generationMessageSchema>;
export type ToolSchemaV1 = z.infer<typeof toolSchemaSchema>;
export type StructuredOutputSpecV1 = z.infer<typeof structuredOutputSpecSchema>;
export type ProviderCapabilitiesV1 = z.infer<typeof providerCapabilitiesSchema>;
export type ModelConfigV1 = z.infer<typeof modelConfigSchema>;
export type ProviderUsageV1 = z.infer<typeof providerUsageSchema>;
export type ProviderErrorV1 = z.infer<typeof providerErrorSchema>;
export type RecordedResponseKeyV1 = z.infer<typeof recordedResponseKeySchema>;
export type GenerationRequestV1 = z.infer<typeof generationRequestSchema>;
export type GenerationResponseV1 = z.infer<typeof generationResponseSchema>;
export type GenerationArtifactV1 = z.infer<typeof generationArtifactSchema>;
export type ProviderGenerateRequestV1 = z.infer<typeof providerGenerateRequestSchema>;
export type ProviderGenerateResponseV1 = z.infer<typeof providerGenerateResponseSchema>;

File: packages/contracts-python/src/aegis_contracts/generation.py
"""Phase 18 model-provider generation contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from aegis_contracts.errors import ContractErrorCode, ContractValidationError
from aegis_contracts.primitives import CorrelationId, GenerationRequestId, TraceId, UtcTimestamp
from aegis_contracts.versioning import (
    assert_supported_schema_version,
)


class GenerationMessageRole(StrEnum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


class ProviderCapability(StrEnum):
    STRUCTURED_OUTPUT = "structured_output"
    TOOLS = "tools"
    VISION = "vision"
    STREAMING = "streaming"


class ProviderFinishReason(StrEnum):
    STOP = "stop"
    LENGTH = "length"
    TOOL_CALLS = "tool_calls"
    CONTENT_FILTER = "content_filter"
    ERROR = "error"


class ProviderErrorCode(StrEnum):
    VALIDATION_FAILED = "VALIDATION_FAILED"
    CAPABILITY_UNSUPPORTED = "CAPABILITY_UNSUPPORTED"
    CREDENTIALS_MISSING = "CREDENTIALS_MISSING"
    OUTPUT_LIMIT_EXCEEDED = "OUTPUT_LIMIT_EXCEEDED"
    STRUCTURED_OUTPUT_INVALID = "STRUCTURED_OUTPUT_INVALID"
    TIMEOUT = "TIMEOUT"
    RETRY_EXHAUSTED = "RETRY_EXHAUSTED"
    CIRCUIT_OPEN = "CIRCUIT_OPEN"
    PROVIDER_UNAVAILABLE = "PROVIDER_UNAVAILABLE"
    CANCELLATION = "CANCELLATION"
    INTERNAL = "INTERNAL"


class GenerationMessageV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    role: GenerationMessageRole
    content: str = Field(min_length=0, max_length=65536)
    name: str | None = Field(default=None, max_length=128)


class ToolSchemaV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(min_length=1, max_length=2048)
    parameters: dict[str, Any] = Field(default_factory=dict)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("tool_schema", value)
        return value


class StructuredOutputSpecV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    json_schema: dict[str, Any] = Field(alias="jsonSchema")
    strict: bool = True
    max_repair_attempts: int = Field(alias="maxRepairAttempts", ge=0, le=2, default=1)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("structured_output_spec", value)
        return value


class ProviderCapabilitiesV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    capabilities: list[ProviderCapability] = Field(default_factory=list)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("provider_capabilities", value)
        return value


class ModelConfigV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    provider_id: str = Field(alias="providerId", min_length=1, max_length=64)
    model_id: str = Field(alias="modelId", min_length=1, max_length=128)
    prompt_version: str = Field(alias="promptVersion", min_length=1, max_length=64)
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_output_tokens: int | None = Field(default=None, alias="maxOutputTokens", ge=1, le=65536)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("model_config", value)
        return value


class ProviderUsageV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    prompt_tokens: int = Field(alias="promptTokens", ge=0)
    completion_tokens: int = Field(alias="completionTokens", ge=0)
    total_tokens: int = Field(alias="totalTokens", ge=0)
    estimated_cost_usd: float | None = Field(default=None, alias="estimatedCostUsd", ge=0.0)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("provider_usage", value)
        return value

    @model_validator(mode="after")
    def validate_totals(self) -> ProviderUsageV1:
        if self.total_tokens != self.prompt_tokens + self.completion_tokens:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="totalTokens must equal promptTokens + completionTokens",
                details={
                    "promptTokens": self.prompt_tokens,
                    "completionTokens": self.completion_tokens,
                    "totalTokens": self.total_tokens,
                },
            )
        return self


class ProviderErrorV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    code: ProviderErrorCode
    message: str = Field(min_length=1, max_length=2048)
    retryable: bool = False
    details: dict[str, Any] = Field(default_factory=dict)
    trace_id: TraceId | None = Field(default=None, alias="traceId")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("provider_error", value)
        return value


class RecordedResponseKeyV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    provider_id: str = Field(alias="providerId", min_length=1, max_length=64)
    model_id: str = Field(alias="modelId", min_length=1, max_length=128)
    prompt_version: str = Field(alias="promptVersion", min_length=1, max_length=64)
    request_hash: str = Field(alias="requestHash", min_length=8, max_length=128)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("recorded_response_key", value)
        return value


class GenerationRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    request_id: GenerationRequestId = Field(alias="requestId")
    trace_id: TraceId = Field(alias="traceId")
    correlation_id: CorrelationId | None = Field(default=None, alias="correlationId")
    provider_id: str | None = Field(default=None, alias="providerId", max_length=64)
    model_config_ref: ModelConfigV1 = Field(alias="modelConfig")
    messages: list[GenerationMessageV1] = Field(min_length=1, max_length=64)
    tools: list[ToolSchemaV1] = Field(default_factory=list, max_length=32)
    structured_output: StructuredOutputSpecV1 | None = Field(
        default=None, alias="structuredOutput"
    )
    capabilities_required: list[ProviderCapability] = Field(
        default_factory=list, alias="capabilitiesRequired"
    )
    max_output_tokens: int = Field(alias="maxOutputTokens", ge=1, le=65536, default=4096)
    timeout_ms: int | None = Field(default=None, alias="timeoutMs", ge=100, le=600000)
    idempotency_key: str | None = Field(default=None, alias="idempotencyKey", max_length=256)

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("generation_request", value)
        return value


class GenerationResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    request_id: GenerationRequestId = Field(alias="requestId")
    trace_id: TraceId = Field(alias="traceId")
    provider_id: str = Field(alias="providerId", min_length=1, max_length=64)
    model_id: str = Field(alias="modelId", min_length=1, max_length=128)
    prompt_version: str = Field(alias="promptVersion", min_length=1, max_length=64)
    content: str | None = Field(default=None, max_length=65536)
    structured_data: dict[str, Any] | None = Field(default=None, alias="structuredData")
    finish_reason: ProviderFinishReason = Field(alias="finishReason")
    usage: ProviderUsageV1
    latency_ms: int = Field(alias="latencyMs", ge=0)
    artifact_ref: str | None = Field(default=None, alias="artifactRef", max_length=512)
    completed_at: UtcTimestamp = Field(alias="completedAt")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("generation_response", value)
        return value


class GenerationArtifactV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    request_id: GenerationRequestId = Field(alias="requestId")
    trace_id: TraceId = Field(alias="traceId")
    provider_id: str = Field(alias="providerId", min_length=1, max_length=64)
    model_id: str = Field(alias="modelId", min_length=1, max_length=128)
    prompt_version: str = Field(alias="promptVersion", min_length=1, max_length=64)
    latency_ms: int = Field(alias="latencyMs", ge=0)
    usage: ProviderUsageV1 | None = None
    error: ProviderErrorV1 | None = None
    sanitized_request: dict[str, Any] = Field(alias="sanitizedRequest")
    sanitized_response: dict[str, Any] | None = Field(default=None, alias="sanitizedResponse")
    object_storage_ref: str | None = Field(default=None, alias="objectStorageRef", max_length=512)
    recorded_at: UtcTimestamp = Field(alias="recordedAt")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("generation_artifact", value)
        return value


class ProviderGenerateRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    request: GenerationRequestV1
    provider_id: str | None = Field(default=None, alias="providerId", max_length=64)
    dry_run: bool = Field(default=False, alias="dryRun")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("provider_generate_request", value)
        return value


class ProviderGenerateResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    response: GenerationResponseV1 | None = None
    error: ProviderErrorV1 | None = None

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("provider_generate_response", value)
        return value

    @model_validator(mode="after")
    def validate_result(self) -> ProviderGenerateResponseV1:
        if (self.response is None) == (self.error is None):
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="Exactly one of response or error must be set",
            )
        return self
